jsaddle/src/Language/Javascript/JSaddle/Run.hs

```diff
@@ -266,9 +261,17 @@ runJavaScriptInt sendReqsTimeout pendingReqsLimit sendReqsBatch = do
                     `catchError` (\v -> unsafeInlineLiftIO $
                                    putStrLn "JavaScriptException happened in sync callback" >> throwIO (JavaScriptException v))
                 exitSyncFrame myDepth result
-              yield
+              yieldRequests
             Nothing -> error $ "sync callback " <> show callbackId <> " called, but does not exist"
-        SyncCommand_Continue -> yield
+        SyncCommand_Continue -> go
+          where
+            go = do
+              -- Only yieldResults in Continue. While doing StartCallback there could be pending requests
+              -- on JS side which could result in inconsistencies in the state of JS and Haskell
+              results <- yieldResults
+              case results of
+                [] -> either id id <$> race yieldRequests (threadDelay 100 >> go)
```

Comment:
Does the go loop need to include yieldRequests ? That is only the top level needs to call yieldRequests, which blocks, while after that we just want to poll yieldResults IIUC. This avoids lots of `takeMVar`'s blocking (although it might not be as costly as I'm thinking).

Reply:
This is a not nice, but functionally correct logic. Ideally I should not need `threadDelay` and implement it entirely using locks while keeping the semantics same as this. (and perhaps even better is to append `yieldRequests` to `results` if it has data). I will try to re-write this today.

Reply:
It took me a while to (hopefully) get this right. Now there is no polling, just wait on a single MVar.